In [1]:
!pip install neuralforecast

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 170.6/170.6 kB 4.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 776.9/776.9 kB 13.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.5/62.5 MB 12.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 409.6/409.6 kB 43.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 805.2/805.2 kB 69.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 101.7/101.7 kB 15.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 226.8/226.8 kB 30.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.6/78.6 kB 13.1 MB/s eta 0:00:00


In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from neuralforecast import NeuralForecast
from neuralforecast.models import LSTM
from neuralforecast.losses.pytorch import MSE, MAE, MQLoss, DistributionLoss
from neuralforecast.tsdataset import TimeSeriesDataset, TimeSeriesLoader
from numpy.random import seed
from random import randrange
from neuralforecast.losses.numpy import rmse, mape
from pandas.tseries.holiday import USFederalHolidayCalendar
from pandas.tseries.offsets import CustomBusinessDay

In [4]:
df = pd.read_excel("SP100_vol.xlsx", index_col="Date",parse_dates=True)
columns=df.columns
df_static=pd.read_excel("Static_df.xlsx")
n_series=len(df.columns)
train = df.iloc[:int(len(df['NKE'])*0.7)]
validation_length=int(len(df['NKE'])*0.7)-int(len(df['NKE'])*0.5)
train=train.unstack().reset_index(name="Value").rename(columns={"level_0":"unique_id", "Date":"ds", "Value":"y"})

# Hyperparameter Search

In [5]:
n_inputs = [3,5,10,15,21,42]
encoder_layers=[1,2,4,6]
dencoder_layers=[1,2,4,6]
encoder_hidden_size=[50,100,200,300,400,500]
dencoder_hidden_size=[50,100,200,300,400,500]
encoder_dropout=[0,0.2,0.3,0.4]
epochs=[50,100,150,250,350,450,550,650,750]
learning_rate=[0.0005,0.0001,0.00005,0.00001]
num_lr_decays=[5,3,2,1]
scaler_type=["robust","standard",'minmax']
context_size=[5,10,15,20,30,40]
losses=[MSE(),MAE(),MQLoss(level=[80, 90]),DistributionLoss(distribution='StudentT', level=[80, 90])]


In [5]:
for z in range (5):
  i=randrange(6)
  h=randrange(4)
  j=randrange(6)
  j2=randrange(6)
  j3=randrange(6)
  a=randrange(4)
  b=randrange(4)
  k=randrange(9)
  l=randrange(4)
  m=randrange(4)
  n=randrange(4)
  o=randrange(3)
  model=LSTM (h=1, input_size=n_inputs[i],
       encoder_n_layers=encoder_layers[h], encoder_hidden_size=encoder_hidden_size[j],
       encoder_dropout=encoder_dropout[b],
       context_size=context_size[j3], decoder_hidden_size=dencoder_hidden_size[j2],
       decoder_layers=dencoder_layers[a],
       stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
       loss=losses[m],
       max_steps=epochs[k], learning_rate=learning_rate[l],
       num_lr_decays=num_lr_decays[l],
       val_check_steps=500, scaler_type=scaler_type[o],
       random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "LSTM-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["LSTM"].values
  else:
    Y_hat=forecasts["LSTM-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | Encoder Layers: {encoder_layers[h]} | Loss: {m} | Scaler Type: {scaler_type[o]}")
  print(f"Learning Rate: {l} | Encoder Dropout: {encoder_dropout[b]} | Dencoder Layers: {dencoder_layers[a]}")
  print(f"Encoder Hidden Size: {encoder_hidden_size[j]} | Decoder Hidden Size: {dencoder_hidden_size[j2]} | Context Size {context_size[j3]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 127090240


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 102741770


Epoch: 250 | Inputs: 10 | Encoder Layers: 6 | Loss: 0 | Scaler Type: standard
Learning Rate: 1 | Encoder Dropout: 0.4 | Dencoder Layers: 6
Encoder Hidden Size: 100 | Decoder Hidden Size: 400 | Context Size 10
RMSE: 0.006515788542662812 | QLIKE: 0.12547324240973165 | MAE: 0.005460329040670946 | MAPE: 0.6664528825587447


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 101132883


Epoch: 450 | Inputs: 42 | Encoder Layers: 6 | Loss: 1 | Scaler Type: standard
Learning Rate: 0 | Encoder Dropout: 0 | Dencoder Layers: 2
Encoder Hidden Size: 300 | Decoder Hidden Size: 400 | Context Size 15
RMSE: 0.0029296027587516647 | QLIKE: 0.031740919959667144 | MAE: 0.0019283695351540657 | MAPE: 0.20105013044096096


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 95679155


Epoch: 50 | Inputs: 42 | Encoder Layers: 4 | Loss: 2 | Scaler Type: minmax
Learning Rate: 2 | Encoder Dropout: 0.4 | Dencoder Layers: 4
Encoder Hidden Size: 50 | Decoder Hidden Size: 200 | Context Size 10
RMSE: 0.008039789839818652 | QLIKE: 1.1838845615605176 | MAE: 0.006789361127242138 | MAPE: 0.641091223491046


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 9087469


Epoch: 650 | Inputs: 21 | Encoder Layers: 1 | Loss: 3 | Scaler Type: standard
Learning Rate: 2 | Encoder Dropout: 0.3 | Dencoder Layers: 4
Encoder Hidden Size: 200 | Decoder Hidden Size: 200 | Context Size 20
RMSE: 0.0032601809200851245 | QLIKE: 0.03682976055533428 | MAE: 0.002093875246046763 | MAPE: 0.20728156002790446


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 250 | Inputs: 42 | Encoder Layers: 2 | Loss: 2 | Scaler Type: robust
Learning Rate: 0 | Encoder Dropout: 0 | Dencoder Layers: 1
Encoder Hidden Size: 300 | Decoder Hidden Size: 50 | Context Size 30
RMSE: 0.00491807938778896 | QLIKE: 0.0793694114706802 | MAE: 0.003623722181811774 | MAPE: 0.41221405799417526


In [6]:
for z in range (5):
  i=randrange(6)
  h=randrange(4)
  j=randrange(6)
  j2=randrange(6)
  j3=randrange(6)
  a=randrange(4)
  b=randrange(4)
  k=randrange(9)
  l=randrange(4)
  m=randrange(4)
  n=randrange(4)
  o=randrange(3)
  model=LSTM (h=1, input_size=n_inputs[i],
       encoder_n_layers=encoder_layers[h], encoder_hidden_size=encoder_hidden_size[j],
       encoder_dropout=encoder_dropout[b],
       context_size=context_size[j3], decoder_hidden_size=dencoder_hidden_size[j2],
       decoder_layers=dencoder_layers[a],
       stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
       loss=losses[m],
       max_steps=epochs[k], learning_rate=learning_rate[l],
       num_lr_decays=num_lr_decays[l],
       val_check_steps=500, scaler_type=scaler_type[o],
       random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "LSTM-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["LSTM"].values
  else:
    Y_hat=forecasts["LSTM-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | Encoder Layers: {encoder_layers[h]} | Loss: {m} | Scaler Type: {scaler_type[o]}")
  print(f"Learning Rate: {l} | Encoder Dropout: {encoder_dropout[b]} | Dencoder Layers: {dencoder_layers[a]}")
  print(f"Encoder Hidden Size: {encoder_hidden_size[j]} | Decoder Hidden Size: {dencoder_hidden_size[j2]} | Context Size {context_size[j3]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 42580966


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 112861500


Epoch: 50 | Inputs: 10 | Encoder Layers: 2 | Loss: 2 | Scaler Type: standard
Learning Rate: 1 | Encoder Dropout: 0.3 | Dencoder Layers: 4
Encoder Hidden Size: 500 | Decoder Hidden Size: 200 | Context Size 5
RMSE: 0.005789126402608702 | QLIKE: 0.10698545602112412 | MAE: 0.0047556528525672 | MAPE: 0.5747462751582153


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 38659600


Epoch: 550 | Inputs: 15 | Encoder Layers: 1 | Loss: 0 | Scaler Type: minmax
Learning Rate: 1 | Encoder Dropout: 0.3 | Dencoder Layers: 2
Encoder Hidden Size: 500 | Decoder Hidden Size: 100 | Context Size 30
RMSE: 0.19274666827856715 | QLIKE: 2.016187957176467 | MAE: 0.1793312921195895 | MAPE: 20.42409131727743


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 86305696


Epoch: 650 | Inputs: 21 | Encoder Layers: 1 | Loss: 3 | Scaler Type: minmax
Learning Rate: 3 | Encoder Dropout: 0 | Dencoder Layers: 4
Encoder Hidden Size: 300 | Decoder Hidden Size: 500 | Context Size 40
RMSE: 0.008466528212745869 | QLIKE: 0.16366856547373784 | MAE: 0.0069028335269970525 | MAPE: 0.8444321408480305


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 7113306


Epoch: 250 | Inputs: 3 | Encoder Layers: 6 | Loss: 3 | Scaler Type: minmax
Learning Rate: 0 | Encoder Dropout: 0 | Dencoder Layers: 6
Encoder Hidden Size: 100 | Decoder Hidden Size: 500 | Context Size 15
RMSE: 0.0059957713082242564 | QLIKE: 0.10454651132631537 | MAE: 0.004612312568092716 | MAPE: 0.5484839724223292


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 750 | Inputs: 21 | Encoder Layers: 6 | Loss: 1 | Scaler Type: minmax
Learning Rate: 0 | Encoder Dropout: 0.4 | Dencoder Layers: 6
Encoder Hidden Size: 400 | Decoder Hidden Size: 500 | Context Size 30
RMSE: 0.005472353129784601 | QLIKE: 0.09289943847269554 | MAE: 0.004121697841090206 | MAPE: 0.47978804130813607


In [7]:
for z in range (5):
  i=randrange(6)
  h=randrange(4)
  j=randrange(6)
  j2=randrange(6)
  j3=randrange(6)
  a=randrange(4)
  b=randrange(4)
  k=randrange(9)
  l=randrange(4)
  m=randrange(4)
  n=randrange(4)
  o=randrange(3)
  model=LSTM (h=1, input_size=n_inputs[i],
       encoder_n_layers=encoder_layers[h], encoder_hidden_size=encoder_hidden_size[j],
       encoder_dropout=encoder_dropout[b],
       context_size=context_size[j3], decoder_hidden_size=dencoder_hidden_size[j2],
       decoder_layers=dencoder_layers[a],
       stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
       loss=losses[m],
       max_steps=epochs[k], learning_rate=learning_rate[l],
       num_lr_decays=num_lr_decays[l],
       val_check_steps=500, scaler_type=scaler_type[o],
       random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "LSTM-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["LSTM"].values
  else:
    Y_hat=forecasts["LSTM-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | Encoder Layers: {encoder_layers[h]} | Loss: {m} | Scaler Type: {scaler_type[o]}")
  print(f"Learning Rate: {l} | Encoder Dropout: {encoder_dropout[b]} | Dencoder Layers: {dencoder_layers[a]}")
  print(f"Encoder Hidden Size: {encoder_hidden_size[j]} | Decoder Hidden Size: {dencoder_hidden_size[j2]} | Context Size {context_size[j3]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 27810397


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 107033145


Epoch: 750 | Inputs: 3 | Encoder Layers: 1 | Loss: 3 | Scaler Type: standard
Learning Rate: 2 | Encoder Dropout: 0.4 | Dencoder Layers: 4
Encoder Hidden Size: 100 | Decoder Hidden Size: 400 | Context Size 20
RMSE: 0.003967567754173722 | QLIKE: 0.08092482520919853 | MAE: 0.002755524670511434 | MAPE: 0.25718039469727344


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 15173153


Epoch: 550 | Inputs: 3 | Encoder Layers: 2 | Loss: 1 | Scaler Type: robust
Learning Rate: 3 | Encoder Dropout: 0.2 | Dencoder Layers: 6
Encoder Hidden Size: 100 | Decoder Hidden Size: 500 | Context Size 10
RMSE: 0.004892366286568713 | QLIKE: 0.0837510109716444 | MAE: 0.003833854041594992 | MAPE: 0.4488505441811497


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 89688495


Epoch: 450 | Inputs: 15 | Encoder Layers: 1 | Loss: 0 | Scaler Type: standard
Learning Rate: 2 | Encoder Dropout: 0.2 | Dencoder Layers: 1
Encoder Hidden Size: 300 | Decoder Hidden Size: 500 | Context Size 20
RMSE: 0.027628063827359026 | QLIKE: 0.599704628485438 | MAE: 0.025574558305683788 | MAPE: 2.9857629860157453


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 47052152


Epoch: 100 | Inputs: 5 | Encoder Layers: 2 | Loss: 2 | Scaler Type: standard
Learning Rate: 3 | Encoder Dropout: 0.2 | Dencoder Layers: 1
Encoder Hidden Size: 300 | Decoder Hidden Size: 50 | Context Size 5
RMSE: 0.006416054199405341 | QLIKE: 0.12293599538191638 | MAE: 0.005365051109631944 | MAPE: 0.6542199257984066


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 250 | Inputs: 3 | Encoder Layers: 4 | Loss: 2 | Scaler Type: standard
Learning Rate: 2 | Encoder Dropout: 0 | Dencoder Layers: 6
Encoder Hidden Size: 400 | Decoder Hidden Size: 500 | Context Size 10
RMSE: 0.004623001657402837 | QLIKE: 0.09651618136058937 | MAE: 0.003017195196242025 | MAPE: 0.2749898057475325


In [8]:
for z in range (5):
  i=randrange(6)
  h=randrange(4)
  j=randrange(6)
  j2=randrange(6)
  j3=randrange(6)
  a=randrange(4)
  b=randrange(4)
  k=randrange(9)
  l=randrange(4)
  m=randrange(4)
  n=randrange(4)
  o=randrange(3)
  model=LSTM (h=1, input_size=n_inputs[i],
       encoder_n_layers=encoder_layers[h], encoder_hidden_size=encoder_hidden_size[j],
       encoder_dropout=encoder_dropout[b],
       context_size=context_size[j3], decoder_hidden_size=dencoder_hidden_size[j2],
       decoder_layers=dencoder_layers[a],
       stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
       loss=losses[m],
       max_steps=epochs[k], learning_rate=learning_rate[l],
       num_lr_decays=num_lr_decays[l],
       val_check_steps=500, scaler_type=scaler_type[o],
       random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "LSTM-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["LSTM"].values
  else:
    Y_hat=forecasts["LSTM-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | Encoder Layers: {encoder_layers[h]} | Loss: {m} | Scaler Type: {scaler_type[o]}")
  print(f"Learning Rate: {l} | Encoder Dropout: {encoder_dropout[b]} | Dencoder Layers: {dencoder_layers[a]}")
  print(f"Encoder Hidden Size: {encoder_hidden_size[j]} | Decoder Hidden Size: {dencoder_hidden_size[j2]} | Context Size {context_size[j3]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 122986805


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 29505651


Epoch: 350 | Inputs: 42 | Encoder Layers: 2 | Loss: 1 | Scaler Type: standard
Learning Rate: 3 | Encoder Dropout: 0.3 | Dencoder Layers: 2
Encoder Hidden Size: 200 | Decoder Hidden Size: 500 | Context Size 30
RMSE: 0.005753256787505863 | QLIKE: 0.10596133249053086 | MAE: 0.0047369089658984695 | MAPE: 0.5732193050156702


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 25062493


Epoch: 550 | Inputs: 3 | Encoder Layers: 2 | Loss: 3 | Scaler Type: robust
Learning Rate: 0 | Encoder Dropout: 0.3 | Dencoder Layers: 6
Encoder Hidden Size: 50 | Decoder Hidden Size: 50 | Context Size 40
RMSE: 0.004525562206365586 | QLIKE: 0.06837860554144799 | MAE: 0.003438208483121311 | MAPE: 0.39596766798427996


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 40243853


Epoch: 100 | Inputs: 10 | Encoder Layers: 6 | Loss: 0 | Scaler Type: standard
Learning Rate: 1 | Encoder Dropout: 0.2 | Dencoder Layers: 2
Encoder Hidden Size: 100 | Decoder Hidden Size: 50 | Context Size 10
RMSE: 0.006602715875421277 | QLIKE: 0.12768293170484113 | MAE: 0.0055436745683677385 | MAPE: 0.6771162605233187


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 22647648


Epoch: 450 | Inputs: 3 | Encoder Layers: 4 | Loss: 1 | Scaler Type: standard
Learning Rate: 3 | Encoder Dropout: 0.3 | Dencoder Layers: 4
Encoder Hidden Size: 500 | Decoder Hidden Size: 100 | Context Size 10
RMSE: 0.005692862620672563 | QLIKE: 0.10448936769865506 | MAE: 0.004667052983643149 | MAPE: 0.5633103537519294


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 750 | Inputs: 15 | Encoder Layers: 4 | Loss: 1 | Scaler Type: robust
Learning Rate: 1 | Encoder Dropout: 0 | Dencoder Layers: 6
Encoder Hidden Size: 200 | Decoder Hidden Size: 100 | Context Size 15
RMSE: 0.00310481275973814 | QLIKE: 0.03512637542564985 | MAE: 0.0020099280266493444 | MAPE: 0.20414072893803278


In [9]:
for z in range (5):
  i=randrange(6)
  h=randrange(4)
  j=randrange(6)
  j2=randrange(6)
  j3=randrange(6)
  a=randrange(4)
  b=randrange(4)
  k=randrange(9)
  l=randrange(4)
  m=randrange(4)
  n=randrange(4)
  o=randrange(3)
  model=LSTM (h=1, input_size=n_inputs[i],
       encoder_n_layers=encoder_layers[h], encoder_hidden_size=encoder_hidden_size[j],
       encoder_dropout=encoder_dropout[b],
       context_size=context_size[j3], decoder_hidden_size=dencoder_hidden_size[j2],
       decoder_layers=dencoder_layers[a],
       stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
       loss=losses[m],
       max_steps=epochs[k], learning_rate=learning_rate[l],
       num_lr_decays=num_lr_decays[l],
       val_check_steps=500, scaler_type=scaler_type[o],
       random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "LSTM-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["LSTM"].values
  else:
    Y_hat=forecasts["LSTM-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | Encoder Layers: {encoder_layers[h]} | Loss: {m} | Scaler Type: {scaler_type[o]}")
  print(f"Learning Rate: {l} | Encoder Dropout: {encoder_dropout[b]} | Dencoder Layers: {dencoder_layers[a]}")
  print(f"Encoder Hidden Size: {encoder_hidden_size[j]} | Decoder Hidden Size: {dencoder_hidden_size[j2]} | Context Size {context_size[j3]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 103648375


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 65015972


Epoch: 550 | Inputs: 21 | Encoder Layers: 6 | Loss: 0 | Scaler Type: standard
Learning Rate: 0 | Encoder Dropout: 0.3 | Dencoder Layers: 6
Encoder Hidden Size: 500 | Decoder Hidden Size: 50 | Context Size 10
RMSE: 0.0028520471088638744 | QLIKE: 0.03021055821632855 | MAE: 0.0018587305644489087 | MAPE: 0.1950729271343313


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 16162781


Epoch: 250 | Inputs: 15 | Encoder Layers: 1 | Loss: 3 | Scaler Type: minmax
Learning Rate: 3 | Encoder Dropout: 0.3 | Dencoder Layers: 1
Encoder Hidden Size: 50 | Decoder Hidden Size: 300 | Context Size 10
RMSE: 0.005657476417861742 | QLIKE: 0.10017897973137158 | MAE: 0.004152789025743407 | MAPE: 0.4719019051977144


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 60668776


Epoch: 50 | Inputs: 10 | Encoder Layers: 6 | Loss: 2 | Scaler Type: minmax
Learning Rate: 0 | Encoder Dropout: 0 | Dencoder Layers: 4
Encoder Hidden Size: 200 | Decoder Hidden Size: 50 | Context Size 20
RMSE: 0.005910821592873797 | QLIKE: 0.1026929973842769 | MAE: 0.004538027625829061 | MAPE: 0.5384291104942283


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 77953746


Epoch: 50 | Inputs: 15 | Encoder Layers: 1 | Loss: 2 | Scaler Type: robust
Learning Rate: 0 | Encoder Dropout: 0.2 | Dencoder Layers: 1
Encoder Hidden Size: 200 | Decoder Hidden Size: 100 | Context Size 15
RMSE: 0.0050530388385235155 | QLIKE: 0.0878581289165262 | MAE: 0.004016944318541215 | MAPE: 0.4750417478993486


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 100 | Inputs: 10 | Encoder Layers: 2 | Loss: 1 | Scaler Type: standard
Learning Rate: 1 | Encoder Dropout: 0.4 | Dencoder Layers: 6
Encoder Hidden Size: 200 | Decoder Hidden Size: 500 | Context Size 15
RMSE: 0.005578202679054434 | QLIKE: 0.10158476609869768 | MAE: 0.004545896936054993 | MAPE: 0.5468340179512107


In [10]:
for z in range (5):
  i=randrange(6)
  h=randrange(4)
  j=randrange(6)
  j2=randrange(6)
  j3=randrange(6)
  a=randrange(4)
  b=randrange(4)
  k=randrange(9)
  l=randrange(4)
  m=randrange(4)
  n=randrange(4)
  o=randrange(3)
  model=LSTM (h=1, input_size=n_inputs[i],
       encoder_n_layers=encoder_layers[h], encoder_hidden_size=encoder_hidden_size[j],
       encoder_dropout=encoder_dropout[b],
       context_size=context_size[j3], decoder_hidden_size=dencoder_hidden_size[j2],
       decoder_layers=dencoder_layers[a],
       stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
       loss=losses[m],
       max_steps=epochs[k], learning_rate=learning_rate[l],
       num_lr_decays=num_lr_decays[l],
       val_check_steps=500, scaler_type=scaler_type[o],
       random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "LSTM-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["LSTM"].values
  else:
    Y_hat=forecasts["LSTM-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | Encoder Layers: {encoder_layers[h]} | Loss: {m} | Scaler Type: {scaler_type[o]}")
  print(f"Learning Rate: {l} | Encoder Dropout: {encoder_dropout[b]} | Dencoder Layers: {dencoder_layers[a]}")
  print(f"Encoder Hidden Size: {encoder_hidden_size[j]} | Decoder Hidden Size: {dencoder_hidden_size[j2]} | Context Size {context_size[j3]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 64162053


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 76816826


Epoch: 550 | Inputs: 42 | Encoder Layers: 1 | Loss: 3 | Scaler Type: minmax
Learning Rate: 0 | Encoder Dropout: 0.2 | Dencoder Layers: 4
Encoder Hidden Size: 50 | Decoder Hidden Size: 500 | Context Size 30
RMSE: 0.00602094747562565 | QLIKE: 0.11141125023109891 | MAE: 0.005128240288703113 | MAPE: 0.6210826301310997


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 108966904


Epoch: 50 | Inputs: 21 | Encoder Layers: 6 | Loss: 2 | Scaler Type: standard
Learning Rate: 1 | Encoder Dropout: 0.3 | Dencoder Layers: 6
Encoder Hidden Size: 400 | Decoder Hidden Size: 500 | Context Size 5
RMSE: 0.0051320347115373375 | QLIKE: 0.09021957288814014 | MAE: 0.004086030075972803 | MAPE: 0.48384620037720794


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 76554544


Epoch: 750 | Inputs: 21 | Encoder Layers: 4 | Loss: 3 | Scaler Type: minmax
Learning Rate: 1 | Encoder Dropout: 0 | Dencoder Layers: 1
Encoder Hidden Size: 500 | Decoder Hidden Size: 400 | Context Size 5
RMSE: 0.004660309407738931 | QLIKE: 0.07921114565361866 | MAE: 0.0032557038902106075 | MAPE: 0.3412744187653747


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 12914533


Epoch: 150 | Inputs: 5 | Encoder Layers: 1 | Loss: 3 | Scaler Type: robust
Learning Rate: 2 | Encoder Dropout: 0 | Dencoder Layers: 1
Encoder Hidden Size: 400 | Decoder Hidden Size: 300 | Context Size 5
RMSE: 0.004833648494835187 | QLIKE: 0.0819414319498974 | MAE: 0.003796641647653612 | MAPE: 0.44542951965386623


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 350 | Inputs: 3 | Encoder Layers: 4 | Loss: 1 | Scaler Type: standard
Learning Rate: 1 | Encoder Dropout: 0 | Dencoder Layers: 2
Encoder Hidden Size: 500 | Decoder Hidden Size: 500 | Context Size 40
RMSE: 0.04216493115636849 | QLIKE: 0.8423063908812266 | MAE: 0.03924078255338521 | MAPE: 4.537744907130497


In [11]:
for z in range (5):
  i=randrange(6)
  h=randrange(4)
  j=randrange(6)
  j2=randrange(6)
  j3=randrange(6)
  a=randrange(4)
  b=randrange(4)
  k=randrange(9)
  l=randrange(4)
  m=randrange(4)
  n=randrange(4)
  o=randrange(3)
  model=LSTM (h=1, input_size=n_inputs[i],
       encoder_n_layers=encoder_layers[h], encoder_hidden_size=encoder_hidden_size[j],
       encoder_dropout=encoder_dropout[b],
       context_size=context_size[j3], decoder_hidden_size=dencoder_hidden_size[j2],
       decoder_layers=dencoder_layers[a],
       stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
       loss=losses[m],
       max_steps=epochs[k], learning_rate=learning_rate[l],
       num_lr_decays=num_lr_decays[l],
       val_check_steps=500, scaler_type=scaler_type[o],
       random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "LSTM-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["LSTM"].values
  else:
    Y_hat=forecasts["LSTM-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | Encoder Layers: {encoder_layers[h]} | Loss: {m} | Scaler Type: {scaler_type[o]}")
  print(f"Learning Rate: {l} | Encoder Dropout: {encoder_dropout[b]} | Dencoder Layers: {dencoder_layers[a]}")
  print(f"Encoder Hidden Size: {encoder_hidden_size[j]} | Decoder Hidden Size: {dencoder_hidden_size[j2]} | Context Size {context_size[j3]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 106185793


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 65009864


Epoch: 100 | Inputs: 42 | Encoder Layers: 6 | Loss: 3 | Scaler Type: standard
Learning Rate: 3 | Encoder Dropout: 0.3 | Dencoder Layers: 6
Encoder Hidden Size: 200 | Decoder Hidden Size: 500 | Context Size 15
RMSE: 0.006742322732212172 | QLIKE: 0.13113831892554148 | MAE: 0.005672785066412177 | MAPE: 0.6934935915346194


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 31934408


Epoch: 100 | Inputs: 42 | Encoder Layers: 1 | Loss: 0 | Scaler Type: standard
Learning Rate: 2 | Encoder Dropout: 0.2 | Dencoder Layers: 2
Encoder Hidden Size: 50 | Decoder Hidden Size: 100 | Context Size 20
RMSE: 0.006774871000443508 | QLIKE: 0.1320390254053692 | MAE: 0.005718619334385856 | MAPE: 0.6997530563892319


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 96930212


Epoch: 100 | Inputs: 10 | Encoder Layers: 6 | Loss: 0 | Scaler Type: robust
Learning Rate: 1 | Encoder Dropout: 0.2 | Dencoder Layers: 6
Encoder Hidden Size: 300 | Decoder Hidden Size: 500 | Context Size 10
RMSE: 0.0052190646602135035 | QLIKE: 0.09223415225801686 | MAE: 0.004187757519106457 | MAPE: 0.49826640670012784


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 54361467


Epoch: 100 | Inputs: 42 | Encoder Layers: 4 | Loss: 2 | Scaler Type: robust
Learning Rate: 0 | Encoder Dropout: 0.4 | Dencoder Layers: 4
Encoder Hidden Size: 50 | Decoder Hidden Size: 400 | Context Size 40
RMSE: 0.005269879965979929 | QLIKE: 0.0935265093904045 | MAE: 0.004246236835895999 | MAPE: 0.5065780972916563


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 250 | Inputs: 15 | Encoder Layers: 4 | Loss: 3 | Scaler Type: robust
Learning Rate: 0 | Encoder Dropout: 0.2 | Dencoder Layers: 1
Encoder Hidden Size: 50 | Decoder Hidden Size: 400 | Context Size 40
RMSE: 0.003931727276107689 | QLIKE: 0.05631681306219822 | MAE: 0.002655174663917594 | MAPE: 0.2654418004505696


In [12]:
for z in range (5):
  i=randrange(6)
  h=randrange(4)
  j=randrange(6)
  j2=randrange(6)
  j3=randrange(6)
  a=randrange(4)
  b=randrange(4)
  k=randrange(9)
  l=randrange(4)
  m=randrange(4)
  n=randrange(4)
  o=randrange(3)
  model=LSTM (h=1, input_size=n_inputs[i],
       encoder_n_layers=encoder_layers[h], encoder_hidden_size=encoder_hidden_size[j],
       encoder_dropout=encoder_dropout[b],
       context_size=context_size[j3], decoder_hidden_size=dencoder_hidden_size[j2],
       decoder_layers=dencoder_layers[a],
       stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
       loss=losses[m],
       max_steps=epochs[k], learning_rate=learning_rate[l],
       num_lr_decays=num_lr_decays[l],
       val_check_steps=500, scaler_type=scaler_type[o],
       random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "LSTM-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["LSTM"].values
  else:
    Y_hat=forecasts["LSTM-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | Encoder Layers: {encoder_layers[h]} | Loss: {m} | Scaler Type: {scaler_type[o]}")
  print(f"Learning Rate: {l} | Encoder Dropout: {encoder_dropout[b]} | Dencoder Layers: {dencoder_layers[a]}")
  print(f"Encoder Hidden Size: {encoder_hidden_size[j]} | Decoder Hidden Size: {dencoder_hidden_size[j2]} | Context Size {context_size[j3]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 38244568


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 13975274


Epoch: 50 | Inputs: 5 | Encoder Layers: 4 | Loss: 2 | Scaler Type: robust
Learning Rate: 1 | Encoder Dropout: 0 | Dencoder Layers: 4
Encoder Hidden Size: 200 | Decoder Hidden Size: 100 | Context Size 15
RMSE: 0.004944056399864424 | QLIKE: 0.08513070729650916 | MAE: 0.0038895305582579663 | MAPE: 0.45655800366797894


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 78504931


Epoch: 350 | Inputs: 5 | Encoder Layers: 4 | Loss: 3 | Scaler Type: minmax
Learning Rate: 1 | Encoder Dropout: 0.3 | Dencoder Layers: 2
Encoder Hidden Size: 500 | Decoder Hidden Size: 50 | Context Size 20
RMSE: 0.006249678897185407 | QLIKE: 0.11044929079077394 | MAE: 0.004851321397117021 | MAPE: 0.5809418480947743


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 128376678


Epoch: 350 | Inputs: 15 | Encoder Layers: 4 | Loss: 3 | Scaler Type: robust
Learning Rate: 0 | Encoder Dropout: 0.2 | Dencoder Layers: 4
Encoder Hidden Size: 200 | Decoder Hidden Size: 300 | Context Size 30
RMSE: 0.0030010262295565603 | QLIKE: 0.03153326466177955 | MAE: 0.0019081308939873313 | MAPE: 0.19144568749707142


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 9478295


Epoch: 650 | Inputs: 3 | Encoder Layers: 6 | Loss: 0 | Scaler Type: minmax
Learning Rate: 3 | Encoder Dropout: 0.4 | Dencoder Layers: 6
Encoder Hidden Size: 300 | Decoder Hidden Size: 400 | Context Size 5
RMSE: 0.007002152315104246 | QLIKE: 0.12829072861783308 | MAE: 0.005552144780427238 | MAPE: 0.673554423088517


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 100 | Inputs: 5 | Encoder Layers: 4 | Loss: 2 | Scaler Type: robust
Learning Rate: 3 | Encoder Dropout: 0.2 | Dencoder Layers: 2
Encoder Hidden Size: 100 | Decoder Hidden Size: 400 | Context Size 20
RMSE: 0.004881380101926158 | QLIKE: 0.08351675743521447 | MAE: 0.0038189986170455753 | MAPE: 0.44646985619806295


INFO:lightning_fabric.utilities.seed:Seed set to 78504931

Epoch: 350 | Inputs: 15 | Encoder Layers: 4 | Loss: 3 | Scaler Type: robust
Learning Rate: 0 | Encoder Dropout: 0.2 | Dencoder Layers: 4
Encoder Hidden Size: 200 | Decoder Hidden Size: 300 | Context Size 30

RMSE: 0.0030010262295565603 | QLIKE: 0.03153326466177955 | MAE: 0.0019081308939873313 | MAPE: 0.19144568749707142

In [6]:
for z in range (5):
  random_seed=randrange(129228148)
  model=LSTM (h=1, input_size=15,
       encoder_n_layers=4, encoder_hidden_size=200,
       encoder_dropout=0.2,
       context_size=30, decoder_hidden_size=300,
       decoder_layers=4,
       stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
       loss=losses[3],
       max_steps=350, learning_rate=learning_rate[0],
       num_lr_decays=num_lr_decays[0],
       val_check_steps=500, scaler_type='robust',
       random_seed=random_seed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=5,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "LSTM-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["LSTM"].values
  else:
    Y_hat=forecasts["LSTM-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Seed: {random_seed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 47232691


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 38967386


Seed: 47232691
RMSE: 0.003194694763872701 | QLIKE: 0.03638378995912343 | MAE: 0.002066072979308422 | MAPE: 0.2074665662539438


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 62000100


Seed: 38967386
RMSE: 0.002768768038645394 | QLIKE: 0.027663506126739915 | MAE: 0.0017679894757906295 | MAPE: 0.18165419037819833


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 71559552


Seed: 62000100
RMSE: 0.0028020537941793465 | QLIKE: 0.02790791157956786 | MAE: 0.0017715762810195394 | MAPE: 0.17977252004133426


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 4171155


Seed: 71559552
RMSE: 0.0031095713974281973 | QLIKE: 0.034196724174146445 | MAE: 0.002004992822070223 | MAPE: 0.20153425038929831


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Seed: 4171155
RMSE: 0.0030213703133298232 | QLIKE: 0.03344272727125931 | MAE: 0.0020129757249645916 | MAPE: 0.2129865136106683


In [7]:
for z in range (5):
  random_seed=randrange(129228148)
  model=LSTM (h=1, input_size=15,
       encoder_n_layers=4, encoder_hidden_size=200,
       encoder_dropout=0.2,
       context_size=30, decoder_hidden_size=300,
       decoder_layers=4,
       stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
       loss=losses[3],
       max_steps=350, learning_rate=learning_rate[0],
       num_lr_decays=num_lr_decays[0],
       val_check_steps=500, scaler_type='robust',
       random_seed=random_seed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=5,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "LSTM-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["LSTM"].values
  else:
    Y_hat=forecasts["LSTM-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Seed: {random_seed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 55868197


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 88016662


Seed: 55868197
RMSE: 0.0030704026409902377 | QLIKE: 0.03327608578507455 | MAE: 0.0019868213105799135 | MAPE: 0.20339624896275837


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 60002989


Seed: 88016662
RMSE: 0.002963854749418169 | QLIKE: 0.031622393708062475 | MAE: 0.0019183669087441747 | MAPE: 0.19702001835457786


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 65845053


Seed: 60002989
RMSE: 0.0028649227260937036 | QLIKE: 0.02938249688005979 | MAE: 0.001836422306647357 | MAPE: 0.18917687479125006


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 49812810


Seed: 65845053
RMSE: 0.003085541968312208 | QLIKE: 0.034087221738840254 | MAE: 0.002025080582926031 | MAPE: 0.20903264888177128


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Seed: 49812810
RMSE: 0.002907666185796212 | QLIKE: 0.030127871572728355 | MAE: 0.0018433329006215826 | MAPE: 0.1839237747228033


In [8]:
for z in range (5):
  random_seed=randrange(129228148)
  model=LSTM (h=1, input_size=15,
       encoder_n_layers=4, encoder_hidden_size=200,
       encoder_dropout=0.2,
       context_size=30, decoder_hidden_size=300,
       decoder_layers=4,
       stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
       loss=losses[3],
       max_steps=350, learning_rate=learning_rate[0],
       num_lr_decays=num_lr_decays[0],
       val_check_steps=500, scaler_type='robust',
       random_seed=random_seed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=5,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "LSTM-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["LSTM"].values
  else:
    Y_hat=forecasts["LSTM-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Seed: {random_seed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 2948078


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 44876972


Seed: 2948078
RMSE: 0.0030522230257712924 | QLIKE: 0.03384982327473524 | MAE: 0.001974207352155667 | MAPE: 0.20064696888996225


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 45367416


Seed: 44876972
RMSE: 0.0028826339191224105 | QLIKE: 0.03080511955902088 | MAE: 0.0019009772992007224 | MAPE: 0.20049000667195113


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 20754962


Seed: 45367416
RMSE: 0.002964891169564506 | QLIKE: 0.032453541586798924 | MAE: 0.0019468952411208272 | MAPE: 0.20188319274022543


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 17678468


Seed: 20754962
RMSE: 0.0030028773369549402 | QLIKE: 0.032680538379841036 | MAE: 0.0019299176396953214 | MAPE: 0.19447519945240185


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Seed: 17678468
RMSE: 0.002718095248264413 | QLIKE: 0.027244813133627782 | MAE: 0.0017362946891489894 | MAPE: 0.17878834803101837


In [9]:
for z in range (5):
  random_seed=randrange(129228148)
  model=LSTM (h=1, input_size=15,
       encoder_n_layers=4, encoder_hidden_size=200,
       encoder_dropout=0.2,
       context_size=30, decoder_hidden_size=300,
       decoder_layers=4,
       stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
       loss=losses[3],
       max_steps=350, learning_rate=learning_rate[0],
       num_lr_decays=num_lr_decays[0],
       val_check_steps=500, scaler_type='robust',
       random_seed=random_seed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=5,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "LSTM-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["LSTM"].values
  else:
    Y_hat=forecasts["LSTM-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Seed: {random_seed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 97568710


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 47018950


Seed: 97568710
RMSE: 0.0028727997231673343 | QLIKE: 0.029934942177339927 | MAE: 0.0018154691392114777 | MAPE: 0.18013659474483792


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 73483298


Seed: 47018950
RMSE: 0.0029434166906209813 | QLIKE: 0.03047980786943089 | MAE: 0.0018583754614495223 | MAPE: 0.18577835213097585


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 25284737


Seed: 73483298
RMSE: 0.0027941063849719473 | QLIKE: 0.02773386699164004 | MAE: 0.0017592839740195496 | MAPE: 0.17699990370140087


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 79753265


Seed: 25284737
RMSE: 0.00302840325038962 | QLIKE: 0.03284168942911603 | MAE: 0.0019889401930329982 | MAPE: 0.20680997628932005


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Seed: 79753265
RMSE: 0.003115093313822682 | QLIKE: 0.03531256029925205 | MAE: 0.002092591428863863 | MAPE: 0.22185628274197094


# Test

In [10]:
test_length=int(len(df['NKE'])*0.3)
df1=df.unstack().reset_index(name="Value").rename(columns={"level_0":"unique_id", "Date":"ds", "Value":"y"})

In [11]:
h=1

In [12]:
model=LSTM (h=1, input_size=15,
       encoder_n_layers=4, encoder_hidden_size=200,
       encoder_dropout=0.2,
       context_size=30, decoder_hidden_size=300,
       decoder_layers=4,
       stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
       loss=losses[3],
       max_steps=350, learning_rate=learning_rate[0],
       num_lr_decays=num_lr_decays[0],
       val_check_steps=500, scaler_type='robust',
     random_seed=17678468)
fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
forecasts = fcst.cross_validation(df=df1,val_size=h,static_df=df_static,n_windows=None, test_size=test_length-test_length%h,step_size=h)
forecasts = forecasts.dropna()
if "LSTM-median" not in list(forecasts.columns.values):
  Y_hat=forecasts["LSTM"].values
else:
  Y_hat=forecasts["LSTM-median"].values
Y_true=forecasts["y"].values
RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 17678468


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

RMSE: 0.004195003958202034 | QLIKE: 0.03486473475443946 | MAE: 0.0025921352539423665 | MAPE: 0.20078472435630526


In [13]:
df1=forecasts.drop(columns=["cutoff", "y"]).set_index("ds")

In [14]:
df1.pivot(columns="unique_id").to_excel("LSTM forecasts.xlsx")